p2p/src/main/kotlin/maru/p2p/MaruPeerManagerService.kt

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright Consensys Software Inc.
+ *
+ * This file is dual-licensed under either the MIT license or Apache License 2.0.
+ * See the LICENSE-MIT and LICENSE-APACHE files in the repository root for details.
+ *
+ * SPDX-License-Identifier: MIT OR Apache-2.0
+ */
+package maru.p2p
+
+import java.time.Duration
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicBoolean
+import maru.config.P2P
+import maru.p2p.discovery.MaruDiscoveryPeer
+import maru.p2p.discovery.MaruDiscoveryService
+import org.apache.logging.log4j.LogManager
+import org.apache.logging.log4j.Logger
+import org.apache.tuweni.bytes.Bytes
+import tech.pegasys.teku.infrastructure.async.SafeFuture
+import tech.pegasys.teku.networking.p2p.network.P2PNetwork
+import tech.pegasys.teku.networking.p2p.network.PeerHandler
+import tech.pegasys.teku.networking.p2p.peer.DisconnectReason
+import tech.pegasys.teku.networking.p2p.peer.Peer
+
+class MaruPeerManagerService(
+  p2pConfig: P2P,
+) : PeerHandler {
+  private val log: Logger = LogManager.getLogger(this::class.java)
+
+  private val maxPeers = p2pConfig.maxPeers
+  private val currentlySearching = AtomicBoolean(false)
+  private val connectionInProgress = mutableListOf<Bytes>()
+
+  private var discoveryService: MaruDiscoveryService? = null
+  private lateinit var p2pNetwork: P2PNetwork<Peer>
+  private var stopCalled = false
+
+  fun start(
+    discoveryService: MaruDiscoveryService?,
+    p2pNetwork: P2PNetwork<Peer>,
+  ) {
+    this.discoveryService = discoveryService
+    this.p2pNetwork = p2pNetwork
+    searchForPeersUntilMaxReached()
+  }
+
+  fun stop(): SafeFuture<Unit> {
+    stopCalled = true
+    return SafeFuture.completedFuture<Unit>(null)
+  }
+
+  private fun searchForPeersUntilMaxReached() {
+    if (!stopCalled && currentlySearching.compareAndSet(false, true)) {
+      discoveryService?.let { discoveryService ->
+        discoveryService
+          .searchForPeers()
+          .orTimeout(Duration.ofSeconds(30L))
+          .whenComplete { availablePeers, throwable ->
+            log.debug("Finished searching for peers. Found {} peers.", availablePeers.size)
+            if (!stopCalled) {
+              if (throwable != null) {
+                log.debug("Failed to discover peers: {}\n{}", throwable.message, throwable.stackTraceToString())
+                discoveryService.getKnownPeers().forEach { peer ->
```

Comment:
Why are we trying to reconnect to known peers when there's an exception?

Reply:
the exception was thrown during the discovery process. getKnownPeers() returns the already discovered peers. 
Because of the way discv5 works, the searchForNewPeers() does not return the newly found peers. We always have to call getKnownPeers() to get all the discovered peers, including the new ones. 
If searchForNewPeers() is throwing an exception we might already have discovered new peers, so we call getKnownPeers() and try to connect to them.

Reply:
But why do we reconnect to known peers? Shouldn't we just do nothing and connect to newly discovered peers on the next iteration when `searchForPeers` hopefully doesn't throw an exception?

Reply:
Also, what happens when we try to connect to a peer that we're already connected to?